### ISSUE TITLE
Possible Optimization for MethodOverrideMatcher

### ISSUE THREAD

**Issue by tylerbenson** [raphw/byte-buddy#550]
Given the [following method](https://github.com/raphw/byte-buddy/blob/master/byte-buddy-dep/src/main/java/net/bytebuddy/matcher/MethodOverrideMatcher.java#L73-L84):
```java
private boolean matches(MethodDescription target, TypeDefinition typeDefinition) {
        for (MethodDescription methodDescription : typeDefinition.getDeclaredMethods().filter(isVirtual())) {
            if (methodDescription.asSignatureToken().equals(target.asSignatureToken())) {
                if (matcher.matches(typeDefinition.asGenericType())) {
                    return true;
                } else {
                    break;
                }
            }
        }
        return false;
    }
```
Perhaps I'm missing something, but it seems like it might be more efficient to move the inner if check to outside the for loop to avoid iterating through the methods when the type doesn't match?
```java
if (matcher.matches(typeDefinition.asGenericType()))
```

Thoughts?

**Comment by raphw** [comment]
My point of view is: I do not know how expensive the supplied matcher is so I assume that it is quite expensive and avoid calling it as good as I can. On the other side, if someone is really out after performance, those folks will often implement their own `ElementMatcher` anyways.

Did you benchmark the usage?

**Comment by tylerbenson** [comment]
I did not benchmark it.  I just discovered this while debugging a matching bug in my code. I understand your concern about not knowing the expense of the matcher.  I was looking at it from the perspective of `isOverriddenFrom(Class<?> type)` where the matcher is pretty trivial, so it seemed rather silly the way it is.  I hadn't considered more expensive cases.

I will let you decide what should be done, and I'd be ok if you decide to just leave it as is.
Thanks for your consideration.
